src/python/giv_api.py

```diff
@@ -0,0 +1,150 @@
+"""giv_api.py
+========================================
+Python bridge for **GIV.jl** using *juliacall*.
+Place this file in ``src/python`` and import with
+
+```python
+from src.python.giv_api import giv, coef_vector, coef_dataframe
+```"""
+
+# one time julia in the enviroment:
+# export PATH="$HOME/.julia/environments/pyjuliapkg/pyjuliapkg/install/bin:$PATH"
+
+from __future__ import annotations
+from pathlib import Path
+from typing import Any, Optional
+import numpy as np
+import pandas as pd
+from juliacall import Main as jl
+
+# ---------------------------------------------------------------------------
+# Paths & one‑time Julia boot
+# ---------------------------------------------------------------------------
+_PROJECT_DIR = Path(__file__).resolve().parents[2]  # ~/GIV.jl repo root
+_SRC_DIR = _PROJECT_DIR / "src"                       # contains GIV.jl
+_julia_ready = False
+
+
+def _init_julia() -> None:
+    """Boot Julia, activate repo environment, import dependencies (once)."""
+    global _julia_ready
+    if _julia_ready:
+        return
+
+    jl.seval("using Pkg")
+    jl.seval(f'Pkg.activate("{_PROJECT_DIR.as_posix()}")')
+    jl.seval(f'push!(LOAD_PATH, "{_SRC_DIR.as_posix()}")')
+    jl.seval("using DataFrames, StatsModels, Tables, GIV")
+
+    _julia_ready = True
+
+# ---------------------------------------------------------------------------
+# Helper – convert guess dict
+# ---------------------------------------------------------------------------
+
+def _py_to_julia_guess(guess: dict):
+    """Convert Python ``dict`` ➜ Julia ``Dict{String,Float64}``."""
+    return jl.Dict([(str(k), float(v)) for k, v in guess.items()])
+
+# ---------------------------------------------------------------------------
+# Public API
+# ---------------------------------------------------------------------------
+
+def giv(
+    df: pd.DataFrame,
+    formula: str,
+    *,
+    id: str,
+    t: str,
+    weight: Optional[str] = None,
+    **kwargs: Any,
+):
+    """Estimate a GIV model from pandas data."""
+
+    _init_julia()
+
+    # --- Convert inputs ------------------------------------------------------
+    jdf = jl.DataFrame(df)
+    jformula = jl.seval(f"@formula({formula})")
+    jid, jt = jl.Symbol(id), jl.Symbol(t)
+    jweight = jl.Symbol(weight) if weight is not None else jl.nothing
+
+    # --- Keyword tweaks ------------------------------------------------------
+    if isinstance(kwargs.get("algorithm"), str):
+        kwargs["algorithm"] = jl.Symbol(kwargs["algorithm"])
+    if isinstance(kwargs.get("guess"), dict):
+        kwargs["guess"] = _py_to_julia_guess(kwargs["guess"])
+
+    # --- Call Julia ----------------------------------------------------------
+    return jl.giv(jdf, jformula, jid, jt, jweight, **kwargs)
```

Comment:
Instead of returning a Julia object, can you return a `givmodel` class in python, with all the fields to python compatible types?